scc/light_client/provider/server.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// Close closes the Server.
+// Closing an already closed Server has no effect
+func (s *Server) Close() {
+	if s.IsClosed() {
+		return
+	}
+	s.client.Close()
+	s.client = nil
+}
+
+// IsClosed returns true if the Server is closed.
+func (s Server) IsClosed() bool {
+	return s.client == nil
+}
+
+// GetCommitteeCertificates returns up to `maxResults` consecutive committee
+// certificates starting from the given period.
+//
+// Parameters:
+// - first: The starting period for which to retrieve committee certificates.
+// - maxResults: The maximum number of committee certificates to retrieve.
+//
+// Returns:
+//   - []cert.CommitteeCertificate: A slice of committee certificates.
+//   - error: Not nil if the provider failed to obtain the requested certificates.
+func (s Server) GetCommitteeCertificates(first scc.Period, maxResults uint64) ([]cert.CommitteeCertificate, error) {
+	if s.IsClosed() {
+		return nil, fmt.Errorf("no client available")
+	}
+
+	results := []ethapi.CommitteeCertificate{}
+	err := s.client.Call(
+		&results,
+		"sonic_getCommitteeCertificates",
+		fmt.Sprintf("0x%x", first),
+		fmt.Sprintf("0x%x", maxResults),
+	)
+	if err != nil {
+		return nil, err
+	}
+	// if too many certificates are returned, drop the excess
+	if uint64(len(results)) > maxResults {
+		results = results[:maxResults]
+	}
+	certs := make([]cert.CommitteeCertificate, len(results))
+	currentPeriod := first
+	for i, res := range results {
+		if res.Period != uint64(currentPeriod) {
+			return nil, fmt.Errorf("committee certificates out of order")
+		}
+		currentPeriod++
+		certs[i] = res.ToCertificate()
+	}
+	return certs, nil
+}
+
+// GetBlockCertificates returns up to `maxResults` consecutive block
+// certificates starting from the given block number.
+//
+// Parameters:
+//   - number: The starting block number for which to retrieve the block certificate.
+//     Can be LatestBlock to retrieve the latest certificates.
+//   - maxResults: The maximum number of block certificates to retrieve.
+//
+// Returns:
+//   - cert.BlockCertificate: The block certificates for the given block number
+//     and the following blocks.
+//   - error: Not nil if the provider failed to obtain the requested certificates.
+func (s Server) GetBlockCertificates(first idx.Block, maxResults uint64) ([]cert.BlockCertificate, error) {
+	if s.IsClosed() {
+		return nil, fmt.Errorf("no client available")
+	}
+
+	var firstString string
+	if first == LatestBlock {
+		firstString = "latest"
+	} else {
+		firstString = fmt.Sprintf("0x%x", first)
+	}
+	results := []ethapi.BlockCertificate{}
+	err := s.client.Call(
+		&results,
+		"sonic_getBlockCertificates",
+		firstString,
+		fmt.Sprintf("0x%x", maxResults),
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(results) == 0 {
+		return nil, fmt.Errorf("no block certificates found")
+	}
+	// if too many certificates are returned, drop the excess
+	if uint64(len(results)) > maxResults {
+		results = results[:maxResults]
+	}
```

Comment:
nit: I would swap those, since `maxResults` may be set to 0 and it seems like the assumption is that the server might not limit itself to this value as it is assumed here; alternatively, have a quick return if `maxResults==0` at the top of the function.